### PR TITLE
Fixed issue #17889: Setting to correct server time not taken into account when sending email invitations

### PR DESCRIPTION
**Pull request by gabrieljenik**

### CODE REVIEW

**Comment by gabrieljenik** [comment]
Tested it. Pending to be merged.

File: application/controllers/admin/Tokens.php [hunk 1]
```diff
@@ -1417,10 +1417,11 @@ public function email(int $iSurveyId)
                     $mail->setFrom(Yii::app()->request->getPost('from_' . $emrow['language']));
                     $mail->rawSubject = $sSubject[$emrow['language']];
                     $mail->rawBody = $sMessage[$emrow['language']];
-                    if (!App()->request->getPost('bypassdatecontrol') == '1' && trim($emrow['validfrom']) != '' && convertDateTimeFormat($emrow['validfrom'], 'Y-m-d H:i:s', 'U') * 1 > date('U') * 1) {
+                    $now = dateShift(date("Y-m-d H:i:s"), "Y-m-d H:i", Yii::app()->getConfig('timeadjust'));
+                    if (!App()->request->getPost('bypassdatecontrol') == '1' && trim($emrow['validfrom']) != '' && strtotime($emrow['validfrom']) > strtotime($now)) {
```
Comment (Shnoulle) [inline application/controllers/admin/Tokens.php:1421]: `!App()->request->getPost('bypassdatecontrol') == '1'` ???

Better : `!App()->request->getPost('bypassdatecontrol') `or `App()->request->getPost('bypassdatecontrol') != '1'` 

Else can we wrap all in a if : 
```
if (!App()->request->getPost('bypassdatecontrol')) {
    if (trim($emrow['validfrom']) != '' && strtotime($emrow['validfrom']) > strtotime($now)) {
        // action
    } elseif (trim($emrow['validuntil']) != '' && strtotime($emrow['validuntil']) < strtotime($now) {
        // action
    }
}
```

really more understandable

**Review by Shnoulle — CHANGES_REQUESTED** [review]
better test (understandable)

File: application/controllers/admin/Tokens.php [hunk 2]
```diff
@@ -1417,62 +1417,75 @@ public function email(int $iSurveyId)
...
+                        $now = dateShift(date("Y-m-d H:i:s"), "Y-m-d H:i", Yii::app()->getConfig('timeadjust'));
+                        $fieldsarray = [
+                            '{FIRSTNAME}' => $emrow['firstname'],
+                            '{LASTNAME}' => $emrow['lastname'],
+                            '{EMAIL}' => $emrow['email'],
+                        ];
+                        if (trim($emrow['validfrom']) != '' && strtotime($emrow['validfrom']) > strtotime($now)) {
+                            $tokenoutput .= $emrow['tid'] . " " . htmlspecialchars(ReplaceFields(gT("Email to {FIRSTNAME} {LASTNAME} ({EMAIL}) delayed: Access code is not yet valid.", 'unescaped'), $fieldsarray)) . "<br />";
+                            $bInvalidDate = true;
+                            continue;
```
Comment (Shnoulle) [inline application/controllers/admin/Tokens.php:1432]: :+1:

File: application/controllers/admin/Tokens.php [hunk 3]
```diff
@@ -1417,62 +1417,75 @@ public function email(int $iSurveyId)
...
+                        } elseif (trim($emrow['validuntil']) != '' && strtotime($emrow['validuntil']) < strtotime($now)) {
+                            $tokenoutput .= $emrow['tid'] . " " . htmlspecialchars(ReplaceFields(gT("Email to {FIRSTNAME} {LASTNAME} ({EMAIL}) skipped: Access code is not valid anymore.", 'unescaped'), $fieldsarray)) . "<br />";
+                            $bInvalidDate = true;
+                            continue;
+                        }
+                    }
+
+                    // If all checks passed, send the email
```
Comment (Shnoulle) [inline application/controllers/admin/Tokens.php:1440]: Yea ! better than an else !

**Review by Shnoulle — APPROVED** [review]

**Pull request merged**